nav2_collision_monitor/src/polygon.cpp

```diff
@@ -168,8 +169,46 @@ bool Polygon::isShapeSet()
   return true;
 }
 
-void Polygon::updatePolygon()
+bool Polygon::isUsingVelocityPolygonSelector()
 {
+  return !velocity_polygons_.empty();
+}
+
+void Polygon::updatePolygon(const Velocity & cmd_vel_in)
+{
+  if (isUsingVelocityPolygonSelector()) {
```

Comment:
I'm trying to figure out here if we have this master bool `isUsingVelocityPolygonSelector` why then we have the VelocityPolygon inside of the Polygon itself and not be a separate class object that has the base type of Polygon to be used in the application separately (or have a Polygon composed object within the VelocityPolygon class). At the end of this `if` condition, we either `return` from getting a polygon in range or logging a warning that there is none, but we never actually enter the logic of the rest of this function.

Reply:
Can the velocity polygon not leverage the subscriptions to change the footprint values that are implemented below? Or setting in different reference frames?

I think some thought needs to be placed into these other features w.r.t. the velocity polygon addition. I don't think there's any philosophical reason that we cannot have the velocity polygon's polygon be adjustable from the subscription like any other polygon?

Reply:
That is really good question, we've previously touched in the original ticket, I recall. One way of the solution that was to make separate module, publishing velocities for CM node, but this idea (for some reasons, I do not recall, maybe performance) we've finally refused. Your idea to make `VelocityPolygon` class to be inherited from `Polygon` seems not much crazy for me. We will have two `updatePolygon()` methods that could be generalized, and depending on some conditions (like velocity), could react and update its shape. Great idea, especially in case of if we:
1. Want to have some unified API of CM & its shapes for future re-usage
2. Refuse the number of if-else cases not related to robot behavior in CM main cycle logic (that I don't like myself, but anyway, it is working solution :smile:, that need be moved into behavior handling logic in a foreseen future).

@kaichie, what do you think about it?

Reply:
The other option is to have dynamic polygon setting support within the velocity polygon to set the different velocity polygons to different values. Its less "ooooh nice design" but would be more straight forward. 

If making the VelocityPolygon derive from Polygon would make things cleaner, that's preferable in my book, but I'm not 100% convinced that wouldn't just be a huge headache and still not fully solve the issue.

Reply:
My thought on this are as follows: if VelocityPolygons support subscriptions to change polygons, wouldn't this be the same as using a normal polygon with the subscription topic? Unless the use case involves users preferring to set the polygon using the topic, even if it's a static polygon.

Conceptually, VelocityPolygon is quite different from the Polygon base class as it contains multiple sub-polygons. Even if it is a derived class of Polygon, I believe it still needs the if-else cases, as CM always checks for the topic first, followed by static points, and finally the VelocityPolygon. But yes then we can use `this->` instead. Or maybe I did not get this part correctly?

If I understand correctly, we are also discussing whether we need to derive VelocityPolygon's sub-polygons from the Polygon base class so that it can support the dynamic changing of polygons with subscriptions. It makes sense to derive it from the Polygon class, but there will likely be some code duplication too (for example, the logic to decide whether to use the topic or static points will be duplicated in the VelocityPolygon class itself). I would be happy to draft a version of this if it helps in the discussion.

Reply:
I agree, I think deriving is not the right answer to get this behavior. The interface for the subscriber would have to be different to specify the name to change - else we'd have to create `N` topics where the polygon name was in the topic name. That's doable, but unwieldy. 

I have two separate topics:
- (1) We skip an entire and massive block of code creating a giant switch in a core function
- (2) The code that the skipped block of code represents is not represented in the VelocityPolygon case and might as well be

I think (2) can be added as a service call or topic within the VelocityPolygon code. (1) I think might just need different abstractions (e.g. break into functions or make explicitly a switch statement) to make clear that its a binary operation being done

Reply:
@SteveMacenski, I am not sure what did you mean in the latest message.
However, the following idea:

> The other option is to have dynamic polygon setting support within the velocity polygon to set the different velocity polygons to different values.

as I understand well - is to make a type of shapes, called `VelocityPolygon`. This shape might not be derived from `Polygon`, as you've previously mentioned for concept and convenience issues, but rather to be a separate parallel instance existing along with usual `Polyhons` and `Circles` (but again, not derived from them). Am I correct, or you meant something else?

Reply:
Or do you mean from one `PolygonVelocity` -> to make N sub-polygons of `Polygon` type (and add them into main `polygons_` array); and then enable/disable these sub-polygons depending on robot velocity by using `PolygonVelocity` as a polygons main brain / handler?

Reply:
I'm suggesting that we change nothing about this PR but:
- VelocityPolygon now when it loads the different velocity regime shapes, exposes a service / topic to change them dynamically via a ROS interface
- Inside of `updatePolygon` we break up this function into 2 separate functions in a `switch()` statement to make it clear that there are two, mutually exclusive paths that are done depending on the nature of the polygon. 

@kaichie does that seem straight forward enough? Lets just try to get this in before ROSCon!

Reply:
Yeap sounds good to me. For each of the sub-polygon in the VelocityPolygon, I will expose it with a topic. Default to `/collision_monitor/set_velocity_polygon/<sub-polygon name>`, unless it is specified in the parameter config(`polygon_sub_topic`). I think this part will be quite similar with the one in `Polygon` base class.

Reply:
OK, works for me. That's alot of topics, but it would be nice to be able to use the Polygon message still instead of having to create a custom one to add a name field. 

Maybe naming as `/collision_monitor/<polygon name>/set_polyon`?

Reply:
@SteveMacenski, thank you for the explanation. This also seems good to me: not change too much in existing PR, and make it to look more logical and usable (topics or services-based handling). Let's make next round of review after update then.

Reply:
Added the dynamic update from topic for velocity polygon(quite abit of duplicated code with the polygon base class), and replaced the huge if-else flow with switch case for handling `updatePolygon()`.